## in this task we have a tweets taking about Iphone 6, and the text have a negative and positive feedback 

In [189]:
import pandas as pd

In [190]:
df = pd.read_csv('iphone6-negative.csv')
df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  
0  I've had an IPhone for like 3 years and I've n...  negative  
1        I dont need the new iphone but I want it :(  negative  
2      fuck the iPhone 6s cus I'm not getting one :(  negative  
3  Hopefully will be able to get my iPhone 6s tod...  negative  
4  @clydesdalebank I've just bought a new iPhone ...  negative

## we need to filter out the data and select the important features ( columns ) 

In [191]:
df = df[['Text','Sentiment']]
df.head()

Text Sentiment
0  I've had an IPhone for like 3 years and I've n...  negative
1        I dont need the new iphone but I want it :(  negative
2      fuck the iPhone 6s cus I'm not getting one :(  negative
3  Hopefully will be able to get my iPhone 6s tod...  negative
4  @clydesdalebank I've just bought a new iPhone ...  negative

## lets count the distribution of the two classes

In [192]:
df.Sentiment.value_counts()

negative    501
positive    501
Name: Sentiment, dtype: int64

## luck that the data is well distrubuted, the number of samples in each class is equally 

In [193]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
x  = cv.fit_transform(df.Text)

In [194]:
x.shape

(1002, 3152)

In [195]:
len(df)

1002

## lets convert it into array 

In [197]:
x  = cv.fit_transform(df.Text).toarray()
x

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [198]:
text_data=[]
for sentence in df.Text:
#     print(sentence)
    text_data.append(sentence)
text_data

["I've had an IPhone for like 3 years and I've never done that  :( hate ma life",
 'I dont need the new iphone but I want it :(',
 "fuck the iPhone 6s cus I'm not getting one :(",
 'Hopefully will be able to get my iPhone 6s today :(\n\nPlane gets in at like 11:30 :/',
 "@clydesdalebank I've just bought a new iPhone and I can't use Apple Pay because you don't support it. When will I get to use it? :(",
 "The new iphone is out today but mine doesn't come til next week :( &lt;/3",
 'Also have to work an iPhone launch at t-mobile :(',
 '@O2 I ended up ringin in my break.. i ordered iphone 6s which in now coming to me on Monday and not today :( bit gutted but fingers crossed',
 '@reviewero iPhone 5 was the best iPhone ever. And this makes me feel so nostalgic. :(',
 "@mitsuhiko scared to update to iOS9 as i don't know what that means for my iTunes match subscription :(",
 'There is a high chance that I will not have my iPhone when I get home :( I think I missed the delivery...',
 '@JohnLeg

In [199]:
filtered_words = []
for word in text_data:   ## arabic 
        #if word not in sw:
        #getting back the words into its root
        filtered_words.append((word))
filtered_words

["I've had an IPhone for like 3 years and I've never done that  :( hate ma life",
 'I dont need the new iphone but I want it :(',
 "fuck the iPhone 6s cus I'm not getting one :(",
 'Hopefully will be able to get my iPhone 6s today :(\n\nPlane gets in at like 11:30 :/',
 "@clydesdalebank I've just bought a new iPhone and I can't use Apple Pay because you don't support it. When will I get to use it? :(",
 "The new iphone is out today but mine doesn't come til next week :( &lt;/3",
 'Also have to work an iPhone launch at t-mobile :(',
 '@O2 I ended up ringin in my break.. i ordered iphone 6s which in now coming to me on Monday and not today :( bit gutted but fingers crossed',
 '@reviewero iPhone 5 was the best iPhone ever. And this makes me feel so nostalgic. :(',
 "@mitsuhiko scared to update to iOS9 as i don't know what that means for my iTunes match subscription :(",
 'There is a high chance that I will not have my iPhone when I get home :( I think I missed the delivery...',
 '@JohnLeg

In [200]:
import re

In [201]:
data_filtred=','.join(text_data)
data_filtred

'I\'ve had an IPhone for like 3 years and I\'ve never done that  :( hate ma life,I dont need the new iphone but I want it :(,fuck the iPhone 6s cus I\'m not getting one :(,Hopefully will be able to get my iPhone 6s today :(\n\nPlane gets in at like 11:30 :/,@clydesdalebank I\'ve just bought a new iPhone and I can\'t use Apple Pay because you don\'t support it. When will I get to use it? :(,The new iphone is out today but mine doesn\'t come til next week :( &lt;/3,Also have to work an iPhone launch at t-mobile :(,@O2 I ended up ringin in my break.. i ordered iphone 6s which in now coming to me on Monday and not today :( bit gutted but fingers crossed,@reviewero iPhone 5 was the best iPhone ever. And this makes me feel so nostalgic. :(,@mitsuhiko scared to update to iOS9 as i don\'t know what that means for my iTunes match subscription :(,There is a high chance that I will not have my iPhone when I get home :( I think I missed the delivery...,@JohnLegere . Why is the inventory of iPhone 

In [57]:
list_of_words=re.split('[\w]+',data_filtred)
list_of_words

['',
 "'",
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 "'",
 ' ',
 ' ',
 ' ',
 '  :( ',
 ' ',
 ' ',
 ',',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' :(,',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 "'",
 ' ',
 ' ',
 ' ',
 ' :(,',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' :(\n\n',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ':',
 ' :/,@',
 ' ',
 "'",
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 "'",
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 "'",
 ' ',
 ' ',
 '. ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 '? :(,',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 "'",
 ' ',
 ' ',
 ' ',
 ' ',
 ' :( &',
 ';/',
 ',',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 '-',
 ' :(,@',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 '.. ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' :( ',
 ' ',
 ' ',
 ' ',
 ' ',
 ',@',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 '. ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 '. :(,@',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 ' ',
 

In [58]:
list_of_words=re.split('[\W]+',data_filtred)
list_of_words

['I',
 've',
 'had',
 'an',
 'IPhone',
 'for',
 'like',
 '3',
 'years',
 'and',
 'I',
 've',
 'never',
 'done',
 'that',
 'hate',
 'ma',
 'life',
 'I',
 'dont',
 'need',
 'the',
 'new',
 'iphone',
 'but',
 'I',
 'want',
 'it',
 'fuck',
 'the',
 'iPhone',
 '6s',
 'cus',
 'I',
 'm',
 'not',
 'getting',
 'one',
 'Hopefully',
 'will',
 'be',
 'able',
 'to',
 'get',
 'my',
 'iPhone',
 '6s',
 'today',
 'Plane',
 'gets',
 'in',
 'at',
 'like',
 '11',
 '30',
 'clydesdalebank',
 'I',
 've',
 'just',
 'bought',
 'a',
 'new',
 'iPhone',
 'and',
 'I',
 'can',
 't',
 'use',
 'Apple',
 'Pay',
 'because',
 'you',
 'don',
 't',
 'support',
 'it',
 'When',
 'will',
 'I',
 'get',
 'to',
 'use',
 'it',
 'The',
 'new',
 'iphone',
 'is',
 'out',
 'today',
 'but',
 'mine',
 'doesn',
 't',
 'come',
 'til',
 'next',
 'week',
 'lt',
 '3',
 'Also',
 'have',
 'to',
 'work',
 'an',
 'iPhone',
 'launch',
 'at',
 't',
 'mobile',
 'O2',
 'I',
 'ended',
 'up',
 'ringin',
 'in',
 'my',
 'break',
 'i',
 'ordered',
 'ip

In [59]:
len(list_of_words)

16731

# or we can split the words without patterns matching 

In [60]:
list_of_words=re.split(' ',data_filtred)
list_of_words

["I've",
 'had',
 'an',
 'IPhone',
 'for',
 'like',
 '3',
 'years',
 'and',
 "I've",
 'never',
 'done',
 'that',
 '',
 ':(',
 'hate',
 'ma',
 'life,I',
 'dont',
 'need',
 'the',
 'new',
 'iphone',
 'but',
 'I',
 'want',
 'it',
 ':(,fuck',
 'the',
 'iPhone',
 '6s',
 'cus',
 "I'm",
 'not',
 'getting',
 'one',
 ':(,Hopefully',
 'will',
 'be',
 'able',
 'to',
 'get',
 'my',
 'iPhone',
 '6s',
 'today',
 ':(\n\nPlane',
 'gets',
 'in',
 'at',
 'like',
 '11:30',
 ':/,@clydesdalebank',
 "I've",
 'just',
 'bought',
 'a',
 'new',
 'iPhone',
 'and',
 'I',
 "can't",
 'use',
 'Apple',
 'Pay',
 'because',
 'you',
 "don't",
 'support',
 'it.',
 'When',
 'will',
 'I',
 'get',
 'to',
 'use',
 'it?',
 ':(,The',
 'new',
 'iphone',
 'is',
 'out',
 'today',
 'but',
 'mine',
 "doesn't",
 'come',
 'til',
 'next',
 'week',
 ':(',
 '&lt;/3,Also',
 'have',
 'to',
 'work',
 'an',
 'iPhone',
 'launch',
 'at',
 't-mobile',
 ':(,@O2',
 'I',
 'ended',
 'up',
 'ringin',
 'in',
 'my',
 'break..',
 'i',
 'ordered',
 'ip

In [61]:
len(list_of_words)

15647

In [62]:
x.shape

(1002, 3152)

In [63]:
import re
 
# Declare a dictionary
dict = {}
 
# Method to check whether the word
# exists in dictionary or not
def uniqueWord(Word):
 
    if Word in dict:
 
        # If the word exists in dictionary then
        # simply increase its count
        dict[words] += 1
 
    else:
 
        # If the word does not exists in
        # dictionary update the dictionary
        # and make its count 1
        dict.update({words: 1})

In [64]:
Total_length=0
for words in list_of_words:
        uniqueWord(words)
for elements in dict:
        if dict[elements] == 1:
            print(elements)
            Total_length+=1

ma
life,I
:(,fuck
cus
:(

Plane
gets
11:30
:/,@clydesdalebank
Pay
til
&lt;/3,Also
t-mobile
:(,@O2
ended
ringin
break..
gutted
fingers
crossed,@reviewero
ever.
nostalgic.
:(,@mitsuhiko
scared
iTunes
match
subscription
:(,There
high
chance
delivery...,@JohnLegere
inventory
6sPlus
stock.
corporate
am.
:(,need
fade
talking
studying
:(,@3minutegames
Upgraded
today,
restored
backup...
progress
Lifeline
games
:(,@sprintnews
ever.,@EE
o2
:(,Grrrrrrr
Got
(and
only)
I'd
waited
:(,@PolarGlobal
Beat
H6
Disappointed
:(
https://t.co/lv6Ml03If6,@katoi
simply
picture
building
left!,@flinto
scrolling
flinto
prototype
problems?
#flinto
it?,Woke
early
:(,Annoyed
parcel
drop-off
place
failed.
Daft
allow
pre-orders.
PLEASE
consider
options
(black)
7?!
There's
white
iPhones,
:(,Sick
However
6s!!!,c
trop
sur
sticky
happiest
boy
world,
#loveyouapple
#fanboy
@tim_cook,Just
fire
alarm
goes
evacuate
:(,@NikiAlbon
#bummer
#myluck
#preorder
#illwait,@MyPapoose
Tuesday
Ive
stuff
promised
today.,Takin
loans
:(,@d_tr

Criscelle
Espiritu
iselpin
https://t.co/WNzcZXFhUT,Hi
Becca
Franks
BeccaFranks1
https://t.co/NP0UogSGEs,Hi
KURAAA
stillPRETTY__
https://t.co/XAL9kOG1yv,Hi
90smilk
https://t.co/ynuvA1N0pX,Hi
sarah
taylrslouis
https://t.co/vFZ1kBb8uX,Hi
PettyWop
KingPthang
https://t.co/zbFjdwYzgR,Hi
Lindsey
Hewitt
ArtyLinz
https://t.co/GUO9nYCvL2,Hi
Nyrie
F.
Nynycharmaigne
https://t.co/uTrWkiFskC,Hi
hailey_
haileybarton97
https://t.co/nxuSFcAyRu,Hi
devin
g0ldpapii
https://t.co/lVWOnx5TK4,Hi
Camille
Serenity
blker_the_berri
https://t.co/Cg14qdB5bB,Hi
emmakwilson3
https://t.co/nXNwX3quLF,Hi
chuck
Chuckoren_
https://t.co/05IMTvFWHj,Happy
objects!
me,
shopping
6.
http://t.co/PyFsWvbmdu,I
:)
And
dear
reader,
pink.
"Rose
gold",
yeah,
riiiiight
;),Hi
Carli
carliiha22
https://t.co/ijxbGwOnby,Hi
Skip
1Skip_
https://t.co/p5VAUpya9m,Hi
Martha
Easton
Martha_Noreen
https://t.co/5UVQhKkSOI,Hi
Sean
englishsean_
https://t.co/Fx81S9Niym,Hi
booty.
peyton_fc
https://t.co/PmnXjxX1lO,Hi
ToSeeHer
DollFaceDarlin
https://t.co/2

In [65]:
Total_length

3191

In [66]:
x.shape

(1002, 3152)

## besides the other symboles that maybe exits the number of unique words looks close to each other with the manual Vs the automatic approach 

In [5]:
x

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [69]:
x.shape

(1002, 3152)

In [70]:
len(df)

1002

# Observations this means that we have 1002 records, with 3152 features ( unique words) as we constructed in our slides 

# we need to minmize these kind of features, its too much

# lets take a look inside these features using method called `cv.vocabulary_`

In [67]:
cv.vocabulary_

{'ve': 2897,
 'had': 1184,
 'an': 249,
 'iphone': 1382,
 'for': 1054,
 'like': 1634,
 'years': 3096,
 'and': 252,
 'never': 1915,
 'done': 840,
 'that': 2693,
 'hate': 1216,
 'ma': 1703,
 'life': 1628,
 'dont': 843,
 'need': 1905,
 'the': 2695,
 'new': 1916,
 'but': 516,
 'want': 2958,
 'it': 1401,
 'fuck': 1074,
 '6s': 104,
 'cus': 713,
 'not': 1952,
 'getting': 1110,
 'one': 2010,
 'hopefully': 1271,
 'will': 3012,
 'be': 392,
 'able': 175,
 'to': 2751,
 'get': 1108,
 'my': 1877,
 'today': 2752,
 'plane': 2123,
 'gets': 1109,
 'in': 1346,
 'at': 331,
 '11': 12,
 '30': 52,
 'clydesdalebank': 617,
 'just': 1480,
 'bought': 461,
 'can': 538,
 'use': 2875,
 'apple': 291,
 'pay': 2077,
 'because': 395,
 'you': 3114,
 'don': 839,
 'support': 2608,
 'when': 3000,
 'is': 1393,
 'out': 2041,
 'mine': 1791,
 'doesn': 834,
 'come': 632,
 'til': 2735,
 'next': 1918,
 'week': 2983,
 'lt': 1690,
 'also': 238,
 'have': 1217,
 'work': 3035,
 'launch': 1588,
 'mobile': 1824,
 'o2': 1977,
 'ended': 91

# Observations we can see that there is words occures in each sentence such as

## **an**
## **i** 
## **have** 
## **ma** 
## **but**
## **had** ...etc

# these kind of words that exits to help the sentence in the context meaning are called `stopwords` 

## these kind of words doesnot effect in the text that decide if its hate or positive sentence

## luckely we have a file that have several stop words 

## Stopwords removing

## opening the file that have stop words customised by our self ( Manually method) 

## as we are using NLTK library so we can search on these kind of list on internet it can be found [here](https://gist.github.com/sebleier/554280)

## download it and save it in text file in your directory

## load the stop words
## we make split `\n` becuase each word in a line

In [72]:
sw = open('sw.txt','r').read().split('\n')

## the second method (automatic method) is by loading the stop words form NLTK library and specify the language 

In [73]:
from nltk.corpus import stopwords
sw = stopwords.words('english')

LookupError: 
**********************************************************************
  Resource stopwords not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('stopwords')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/stopwords

  Searched in:
    - 'C:\\Users\\Mohamed Elhaj-Abdou/nltk_data'
    - 'C:\\Users\\Mohamed Elhaj-Abdou\\Anaconda3-1\\nltk_data'
    - 'C:\\Users\\Mohamed Elhaj-Abdou\\Anaconda3-1\\share\\nltk_data'
    - 'C:\\Users\\Mohamed Elhaj-Abdou\\Anaconda3-1\\lib\\nltk_data'
    - 'C:\\Users\\Mohamed Elhaj-Abdou\\AppData\\Roaming\\nltk_data'
    - 'C:\\nltk_data'
    - 'D:\\nltk_data'
    - 'E:\\nltk_data'
**********************************************************************


## if this error occures you need to download the stopwords first using the following command

In [202]:
stopwords.words('english')

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [74]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to C:\Users\Mohamed Elhaj-
[nltk_data]     Abdou\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [75]:
from nltk.corpus import stopwords
sw = stopwords.words('english')

In [76]:
sw

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [79]:
def sw_rm(text):
    filtered_words = []
    # converting the data into the lowercase to compare it with our stopwords
    for word in text.lower().split():
    # if the word inside the stop words remove it   
        if word not in sw:
            filtered_words.append(word)
    return ' '.join(filtered_words)

In [80]:
String="I have had an IPhone for like 3 years and I've"

In [81]:
sw_rm(String)

"iphone like 3 years i've"

In [83]:
df['clean_text'] = df.Text.apply(sw_rm)
df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  \
0  I've had an IPhone for like 3 years and I've n...  negative   
1        I dont need the new iphone but I want it :(  negative   
2      fuck the iPhone 6s cus I'm not getting one :(  negative   
3  Hopefully will be able to get my iPhone 6s tod...  negative   
4  @clydesdalebank I've just bought a new iPhone ...  negative   

                                          clean_text  
0  i've iphone like 3 years i've never done :( ha...  
1                       dont need new iphone want :(  
2              fuck iphone 6s cus i'm getting one :(  
3  hopefully able get iphone 6s today :( plane ge...  
4  @clydesdalebank i've bought new iphone can't u...

In [84]:
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 3079)

In [203]:
3152-3079

73

## still not happy with the number of features we downgraded from `3152` to `3079` we need to remove more 

## DeNoise

## Removing the unwanted symboles that is not needed such as `@` `:)` `..` `'` 

In [85]:
import re
def DeNoise(text):
    return re.sub('[^a-zA-Z0-9 ]+','',text).strip()

In [89]:
DeNoise("I've had an IPhone for like 3 years and I've n.@#@#$$#:)#..	")

'Ive had an IPhone for like 3 years and Ive n'

In [90]:
df['clean_text'] = df.Text.apply(DeNoise)
df['clean_text'] = df.clean_text.apply(sw_rm)

df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  \
0  I've had an IPhone for like 3 years and I've n...  negative   
1        I dont need the new iphone but I want it :(  negative   
2      fuck the iPhone 6s cus I'm not getting one :(  negative   
3  Hopefully will be able to get my iPhone 6s tod...  negative   
4  @clydesdalebank I've just bought a new iPhone ...  negative   

                                          clean_text  
0   ive iphone like 3 years ive never done hate life  
1                          dont need new iphone want  
2                  fuck iphone 6s cus im getting one  
3  hopefully able get iphone 6s today plane gets ...  
4  clydesdalebank ive bought new iphone cant use ...

## still not happy from `3079` to `3069`

In [91]:
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 3069)

## Stemming returning the text into its original verb (ROOT)we can use `PorterStemmer` from NLTK 

## the most popular used `PorterStemmer` for english and `ISRIStemmer` for Arabic

In [92]:
from nltk.stem import PorterStemmer
ps = PorterStemmer()
ps.stem('playing')

'play'

In [93]:
def stemming(text):
    stemmed_words = []
    for word in text.lower().split():
        stemmed_words.append(ps.stem(word))
    return ' '.join(stemmed_words)

In [100]:
stemming("played playing going I've had an IPhone for like 3 years and I've n...	")

"play play go i'v had an iphon for like 3 year and i'v n..."

## removing the symbols and other unwanted chars
## make steaming 
## remove the stop words

In [106]:
df['clean_text'] = df.Text.apply(DeNoise)
df['clean_text'] = df.clean_text.apply(sw_rm)
df['clean_text'] = df.clean_text.apply(stemming)
df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  \
0  I've had an IPhone for like 3 years and I've n...  negative   
1        I dont need the new iphone but I want it :(  negative   
2      fuck the iPhone 6s cus I'm not getting one :(  negative   
3  Hopefully will be able to get my iPhone 6s tod...  negative   
4  @clydesdalebank I've just bought a new iPhone ...  negative   

                                          clean_text  
0     ive iphon like 3 year ive never done hate life  
1                           dont need new iphon want  
2                        fuck iphon 6s cu im get one  
3    hope abl get iphon 6s today plane get like 1130  
4  clydesdalebank ive bought new iphon cant use a...

In [135]:
df['Sentiment'].unique()

array(['negative', 'positive'], dtype=object)

In [137]:
df[df['Sentiment']=='positive']

TweetID            User_ID  \
501   6.470000e+17        Wyatt Smith   
502   6.470000e+17          Topher Li   
503   6.470000e+17  iPhone 6 Giveaway   
504   6.470000e+17  iPhone 6 Giveaway   
505   6.470000e+17  iPhone 6 Giveaway   
...            ...                ...   
997   6.470000e+17  iPhone 6 Giveaway   
998   6.470000e+17  iPhone 6 Giveaway   
999   6.470000e+17  iPhone 6 Giveaway   
1000  6.470000e+17  iPhone 6 Giveaway   
1001  6.470000e+17  iPhone 6 Giveaway   

                                                   Text Sentiment  \
501   Who is ready for iPhone 6s and iPhone 6s Plus?...  positive   
502   @lucy_larkman there are hacks on how to save h...  positive   
503   Hi Rica Cunanan : ricacunanan09 Do u want to g...  positive   
504   Hi Ariane Valentine : mainlyboredom Do u want ...  positive   
505   Hi eastside_crazii : rthompson672 Do u want to...  positive   
...                                                 ...       ...   
997   Hi solihin : muhdsolihin23 Do u want to get FR...  positive   
998   Hi Andi N.  : inlovewithcamd_ Do u want to get...  positive   
999   Hi nia : deluxechaelin Do u want to get FREE i...  positive   
1000  Hi JBM . : TuesdayBleu Do u want to get FREE i...  positive   
1001  Hi Adam Crandall : adamkcrandall Do u want to ...  positive   

                                             clean_text  
501                    readi iphon 6s iphon 6s plu know  
502   lucylarkman hack save heap batteri time iphon ...  
503   hi rica cunanan ricacunanan09 u want get free ...  
504   hi arian valentin mainlyboredom u want get fre...  
505   hi eastsidecrazii rthompson672 u want get free...  
...                                                 ...  
997   hi solihin muhdsolihin23 u want get free iphon...  
998   hi andi n inlovewithcamd u want get free iphon...  
999   hi nia deluxechaelin u want get free iphon 6 p...  
1000  hi jbm tuesdaybleu u want get free iphon 6 ple...  
1001  hi adam crandal adamkcrandal u want get free i...  

[501 rows x 5 columns]

In [107]:
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 2858)

In [110]:
df['clean_text'] = df.Text.apply(DeNoise)
df['clean_text'] = df.clean_text.apply(stemming)
df['clean_text'] = df.clean_text.apply(sw_rm)

df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  \
0  I've had an IPhone for like 3 years and I've n...  negative   
1        I dont need the new iphone but I want it :(  negative   
2      fuck the iPhone 6s cus I'm not getting one :(  negative   
3  Hopefully will be able to get my iPhone 6s tod...  negative   
4  @clydesdalebank I've just bought a new iPhone ...  negative   

                                          clean_text  
0     ive iphon like 3 year ive never done hate life  
1                           dont need new iphon want  
2                        fuck iphon 6s cu im get one  
3    hope abl get iphon 6s today plane get like 1130  
4  clydesdalebank ive bought new iphon cant use a...

In [109]:
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 2858)

In [111]:
df['clean_text'] = df.clean_text.apply(sw_rm)
df['clean_text'] = df.clean_text.apply(stemming)
df['clean_text'] = df.Text.apply(DeNoise)

df.head()

TweetID         User_ID  \
0  6.470000e+17         Rionagh   
1  6.470000e+17   your highness   
2  6.470000e+17  Little Liggins   
3  6.470000e+17             NaN   
4  6.470000e+17      Kenny Tosh   

                                                Text Sentiment  \
0  I've had an IPhone for like 3 years and I've n...  negative   
1        I dont need the new iphone but I want it :(  negative   
2      fuck the iPhone 6s cus I'm not getting one :(  negative   
3  Hopefully will be able to get my iPhone 6s tod...  negative   
4  @clydesdalebank I've just bought a new iPhone ...  negative   

                                          clean_text  
0  Ive had an IPhone for like 3 years and Ive nev...  
1           I dont need the new iphone but I want it  
2          fuck the iPhone 6s cus Im not getting one  
3  Hopefully will be able to get my iPhone 6s tod...  
4  clydesdalebank Ive just bought a new iPhone an...

In [112]:
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 3176)

In [115]:
df['clean_text'] = df.Text.apply(DeNoise)
df['clean_text'] = df.clean_text.apply(sw_rm)
df['clean_text'] = df.clean_text.apply(stemming)

df.head()
x  = cv.fit_transform(df.clean_text).toarray()
x.shape

(1002, 2858)

# the best one is Denoising --> stop words removing --> steaming

## From `3069` to `2858`

In [116]:
3152-2858

294

# We did our best, we will using logistic regression as a classifier 

# splitting the dataset into training and testset 

In [117]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model    import LogisticRegression

x_train,x_test,y_train,y_test = train_test_split(x,df.Sentiment,test_size=0.20,random_state=110)


In [118]:
clf = LogisticRegression()
clf.fit(x_train,y_train)
round(clf.score(x_test,y_test)*100,2)

86.57

In [119]:
test = 'I dont need the new iphone but I want it :('
test = DeNoise(test)
test = stemming(test)
test = sw_rm(test)
test = cv.transform([test]) ## لو عملت فت هيمسح الجدول القديم ويعمل جدول جديد فيه الكلمات دي بس
clf.predict(test)

array(['negative'], dtype=object)

In [147]:
string=df['Text'].iloc[997]
print(string)

Hi solihin : muhdsolihin23 Do u want to get FREE iPhone 6? Please check my bi0 :) Thx https://t.co/cSuWV6cVx6


In [146]:
test = string
test = DeNoise(test)
test = sw_rm(test)
test = stemming(test)
print(test)
## converting the data into vector as you trained the model on it
test = cv.transform([test])

hi solihin muhdsolihin23 u want get free iphon 6 pleas check bi0 thx httpstcocsuwv6cvx6


In [144]:
clf.predict(test)

array(['positive'], dtype=object)

## Conclusion

In [ ]:
import re
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model    import LogisticRegression
from nltk.stem import PorterStemmer

In [60]:
df = pd.read_csv('iphone6-negative.csv')
df = df[['Text','Sentiment']]
cv = CountVectorizer()
ps = PorterStemmer()   ## arabic
sw = open('sw.txt','r').read().split('\n')   ## arabic

def clean(text):
    text = str(text)
    text = re.sub('[^a-zA-Z0-9 ]+','',text).strip()  ## arabic
    filtered_words = []
    for word in text.lower().split():   ## arabic 
        if word not in sw:
            filtered_words.append(ps.stem(word))
    return ' '.join(filtered_words)

df['clean_text'] = df.Text.apply(clean)

x  = cv.fit_transform(df.clean_text).toarray()

x_train,x_test,y_train,y_test = train_test_split(x,df.Sentiment,test_size=0.20,random_state=110)
clf = LogisticRegression()
clf.fit(x_train,y_train)
round(clf.score(x_test,y_test)*100,2)

86.57

In [62]:
test = 'I dont need the new iphone but I want it :('
test = clean(test)
print(test)
test = cv.transform([test]) ## لو عملت فت هيمسح الجدول القديم ويعمل جدول جديد فيه الكلمات دي بس
clf.predict(test)

dont need new iphon want


array(['negative'], dtype=object)

In [214]:
df = pd.read_csv('dataset.csv')
df

الله يشافيك و يعافيك من كل داء وهامة أخي سيد الوكيل  1
0                                              حلوه اوى   1
1                                                  حقا   1
2                                                صح جدا   1
3                                          اه و النعمة   1
4                                      كلامك جميل ورائع   1
...                                                 ...  ..
1973                                  البرنامج فاشل جدا   0
1974  ضيعت 7 دقايق من عمري بمشاهدة ازنخ و اتفه شخص ب...   0
1975   بكتب ع اوراق الشجر يقول موهبتي وانا صغير الل...   0
1976  هذا اللي بالنص من لجنة الحكم مفكر نفسو وديع ال...   0
1977                                        مسخره والله   0

[1978 rows x 2 columns]

In [215]:
df.columns[0]

'الله يشافيك و يعافيك من كل داء وهامة أخي سيد الوكيل'

In [216]:
df=df.rename(columns={df.columns[0]:'Text'})

In [217]:
df=df.rename(columns={df.columns[1]:'class'})

In [218]:
df.head()

Text  class
0          حلوه اوى      1
1              حقا      1
2            صح جدا      1
3      اه و النعمة      1
4  كلامك جميل ورائع      1

In [219]:
from nltk .corpus import stopwords
from nltk.stem import ISRIStemmer
sw=stopwords.words('arabic')
sw

['إذ',
 'إذا',
 'إذما',
 'إذن',
 'أف',
 'أقل',
 'أكثر',
 'ألا',
 'إلا',
 'التي',
 'الذي',
 'الذين',
 'اللاتي',
 'اللائي',
 'اللتان',
 'اللتيا',
 'اللتين',
 'اللذان',
 'اللذين',
 'اللواتي',
 'إلى',
 'إليك',
 'إليكم',
 'إليكما',
 'إليكن',
 'أم',
 'أما',
 'أما',
 'إما',
 'أن',
 'إن',
 'إنا',
 'أنا',
 'أنت',
 'أنتم',
 'أنتما',
 'أنتن',
 'إنما',
 'إنه',
 'أنى',
 'أنى',
 'آه',
 'آها',
 'أو',
 'أولاء',
 'أولئك',
 'أوه',
 'آي',
 'أي',
 'أيها',
 'إي',
 'أين',
 'أين',
 'أينما',
 'إيه',
 'بخ',
 'بس',
 'بعد',
 'بعض',
 'بك',
 'بكم',
 'بكم',
 'بكما',
 'بكن',
 'بل',
 'بلى',
 'بما',
 'بماذا',
 'بمن',
 'بنا',
 'به',
 'بها',
 'بهم',
 'بهما',
 'بهن',
 'بي',
 'بين',
 'بيد',
 'تلك',
 'تلكم',
 'تلكما',
 'ته',
 'تي',
 'تين',
 'تينك',
 'ثم',
 'ثمة',
 'حاشا',
 'حبذا',
 'حتى',
 'حيث',
 'حيثما',
 'حين',
 'خلا',
 'دون',
 'ذا',
 'ذات',
 'ذاك',
 'ذان',
 'ذانك',
 'ذلك',
 'ذلكم',
 'ذلكما',
 'ذلكن',
 'ذه',
 'ذو',
 'ذوا',
 'ذواتا',
 'ذواتي',
 'ذي',
 'ذين',
 'ذينك',
 'ريث',
 'سوف',
 'سوى',
 'شتان',
 'عدا',
 'عسى',
 'عل'

In [220]:
isris=ISRIStemmer()

In [221]:

def clean(text):
    text = str(text)
    text = re.sub('[^ء-ي0-9 ]+','',text).strip()  ## arabic
    filtered_words = []
    for word in text.lower().split():   ## arabic 
        if word not in sw:
            filtered_words.append(ps.stem(word))
    return ' '.join(filtered_words)


In [222]:

df['clean_text'] = df.Text.apply(clean)

df

Text  class  \
0                                              حلوه اوى      1   
1                                                  حقا      1   
2                                                صح جدا      1   
3                                          اه و النعمة      1   
4                                      كلامك جميل ورائع      1   
...                                                 ...    ...   
1973                                  البرنامج فاشل جدا      0   
1974  ضيعت 7 دقايق من عمري بمشاهدة ازنخ و اتفه شخص ب...      0   
1975   بكتب ع اوراق الشجر يقول موهبتي وانا صغير الل...      0   
1976  هذا اللي بالنص من لجنة الحكم مفكر نفسو وديع ال...      0   
1977                                        مسخره والله      0   

                                             clean_text  
0                                              حلوه اوى  
1                                                        
2                                                صح جدا  
3                                             اه النعمة  
4                                      كلامك جميل ورائع  
...                                                 ...  
1973                                  البرنامج فاشل جدا  
1974  ضيعت 7 دقايق عمري بمشاهدة ازنخ اتفه شخص بالحيا...  
1975  بكتب اوراق الشجر يقول موهبتي وانا صغير الله يس...  
1976  اللي بالنص لجنة الحكم مفكر نفسو وديع الصافي شق...  
1977                                        مسخره والله  

[1978 rows x 3 columns]

In [223]:
x  = cv.fit_transform(df.clean_text).toarray()



In [224]:
x

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [225]:
x.shape

(1978, 6797)

In [226]:
from sklearn.feature_extraction.text import TfidfVectorizer
dfidf=TfidfVectorizer()
x=dfidf.fit_transform(df['clean_text']).toarray()
x.shape

(1978, 6797)

In [227]:
x

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [175]:
df.columns

Index(['Text', 'class', 'clean_text'], dtype='object')

In [177]:
x_train,x_test,y_train,y_test = train_test_split(x,df['class'],test_size=0.20,random_state=110)
clf = LogisticRegression()
clf.fit(x_train,y_train)
round(clf.score(x_test,y_test)*100,2)

80.56

In [178]:
df

Text  class  \
0                                              حلوه اوى      1   
1                                                  حقا      1   
2                                                صح جدا      1   
3                                          اه و النعمة      1   
4                                      كلامك جميل ورائع      1   
...                                                 ...    ...   
1973                                  البرنامج فاشل جدا      0   
1974  ضيعت 7 دقايق من عمري بمشاهدة ازنخ و اتفه شخص ب...      0   
1975   بكتب ع اوراق الشجر يقول موهبتي وانا صغير الل...      0   
1976  هذا اللي بالنص من لجنة الحكم مفكر نفسو وديع ال...      0   
1977                                        مسخره والله      0   

                                             clean_text  
0                                              حلوه اوى  
1                                                        
2                                                صح جدا  
3                                             اه النعمة  
4                                      كلامك جميل ورائع  
...                                                 ...  
1973                                  البرنامج فاشل جدا  
1974  ضيعت 7 دقايق عمري بمشاهدة ازنخ اتفه شخص بالحيا...  
1975  بكتب اوراق الشجر يقول موهبتي وانا صغير الله يس...  
1976  اللي بالنص لجنة الحكم مفكر نفسو وديع الصافي شق...  
1977                                        مسخره والله  

[1978 rows x 3 columns]

In [180]:
test = 'جميله اوي'
test = clean(test)
print(test)
test = dfidf.transform([test])
clf.predict(test)

جميله اوي


array([1], dtype=int64)

In [181]:
test = 'سيئ جدا'
test = clean(test)
print(test)
test = dfidf.transform([test])
clf.predict(test)

سيئ جدا


array([0], dtype=int64)

In [182]:
test = 'زي الزفت'
test = clean(test)
print(test)
test = dfidf.transform([test])
clf.predict(test)

زي الزفت


array([0], dtype=int64)

In [185]:
test =  'ضيعت عمري عليها'
test = clean(test)
print(test)
test = dfidf.transform([test])
clf.predict(test)

ضيعت عمري عليها


array([0], dtype=int64)

In [187]:
test =  'روعه المنتج'
test = clean(test)
print(test)
test = dfidf.transform([test])
clf.predict(test)

روعه المنتج


array([1], dtype=int64)